[lldb] Avoid force loading symbol files in statistics collection

dmpots opened the pull request:
This commit modifies the `DebuggerStats::ReportStatistics` implementation to avoid loading symbol files for unloaded symbols. We collect stats on debugger shutdown and without this change it can cause the debugger to hang for a long while on shutdown if they symbols were not previously loaded (e.g. `settings set target.preload-symbols false`).

The implementation is done by adding an optional parameter to `Module::GetSymtab` to control if the corresponding symbol file will be loaded in the same way that can control it for `Module::GetSymbolFile`.

dmpots committed:
Avoid force loading symbol files in statistics collection

This commit modifies the `DebuggerStats::ReportStatistics` implementation
to avoid loading symbol files for unloaded symbols. We collect stats
on debugger shutdown and without this change it can cause the debugger
to hang for a long while on shutdown if they symbols were not previously
loaded (e.g. `settings set target.preload-symbols false`).

The implementation is done by adding an optional parameter to
`Module::GetSymtab` to control if the corresponding symbol file will
be loaded in the same way that can control it for `Module::GetSymbolFile`.

Jlalond reviewed (approved):
I will let Jeffrey and Greg also chime in but this LGTM. My only concern would be testing the did not parse case by looking at a parse time of 0

dmpots committed:
Add unit test for new GetSymtab overload

Add a unit test to show that symbol tables are only
created on demand.

dmpots committed:
Create target without loading dependents

dmpots commented on lldb/test/Shell/Commands/command-statistics-dump.test:
```diff
@@ -0,0 +1,31 @@
+# When we enable symbol preload and dump stats there should be a non-zero
+# time for parsing symbol tables for the main module.
+# RUN: %lldb %t-main.exe \
+# RUN:       -O "settings set plugin.jit-loader.gdb.enable off" \
```
Thanks for the feedback. I updated the test to create the target as suggested.

Originally I had the test include a '-o "run"' flag as well, but that ran into symbol loading issues as you pointed out. The ASAN/TSAN runtime plugins are loaded automatically and there is no way to disable them. Those plugins force symbol loading by looking for some specific symbols in all executable modules.

Is it a reasonable to expect that we should have a way to turn off new plugins that cause the the test to break if they force symbol loading? Or is there a better way to test this feature?

clayborg replied:
It is fine to not run the program and just test that symbols are not loaded when pre-load is off. That will stop all of the normal system runtime plug-ins from being loaded as they only get loaded when the process is launched.

jeffreytan81 commented on lldb/test/Shell/Commands/command-statistics-dump.test:
```diff
@@ -0,0 +1,31 @@
+# RUN: %clang_host -g %S/Inputs/main.c -o %t-main.exe
+
+# When we enable symbol preload and dump stats there should be a non-zero
+# time for parsing symbol tables for the main module.
```
Can you add top level comment explicitly point out that the test tries to only create the main target module without running or loading dependency to avoid any runtime plugin global symbol lookup. This is a key assumption.

dmpots replied:
Added the comment, thanks!

jeffreytan81 reviewed (approved):
Looks good other than one comment.

dmpots committed:
Add comment explaning assumption around symbol loading

clayborg commented on lldb/unittests/Symbol/SymtabTest.cpp:
```diff
@@ -718,3 +720,40 @@ TEST_F(SymtabTest, TestDecodeCStringMaps) {
+  Symtab *module_symtab = module_sp->GetSymtab(/*can_create=*/false);
+  ASSERT_EQ(module_symtab, nullptr);
+
+  // But it should be created on demand.
+  module_symtab = module_sp->GetSymtab(/*can_create=*/true);
+  ASSERT_NE(module_symtab, nullptr);
```
Maybe add a test to verify that the  symtab gets returns if called again with "can_create = false" to make sure this works as expected, and also to verify the symbol table doesn't change? Something like:
```
Symtab *cached_module_symtab = module_sp->GetSymtab(/*can_create=*/false);
ASSERT_EQ(module_symtab, cached_module_symtab);
```

dmpots replied:
Good idea. Updated the test.

dmpots committed:
Extend unit test to show that we cache the symbol table

Once it has been created then passing `can_create=false` should
have no effect.
